[C10D] Decouple PGNCCL desync from dbg dump

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #114625
* #114615
* __->__ #114614
* #114651

Add TORCH_NCCL_DUMP_DEBUG_INFO env to control dumping independently
of desync debug feature.

Currently default to disabled (so no behavior change by default),
but plan to default this to true after validation.

Moves 'sleep for 30 sec' that used to be after desync debug to before
it. In my view sleeping before desync is equivalent since we always
sleep the same duration, and keeps the code simpler this way.

Fixes #114433

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @yf225 @kiukchung @d4l3k @lucasllc

wconstab committed:
[C10D] Decouple PGNCCL desync from dbg dump

Add TORCH_NCCL_DUMP_DEBUG_INFO env to control dumping independently
of desync debug feature.

Currently default to disabled (so no behavior change by default),
but plan to default this to true after validation.

Moves 'sleep for 30 sec' that used to be after desync debug to before
it. In my view sleeping before desync is equivalent since we always
sleep the same duration, and keeps the code simpler this way.

Fixes #114433

[ghstack-poisoned]

zdevito commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1255,23 +1257,32 @@ void ProcessGroupNCCL::workCleanupLoop() {
+              }
+            }
+
+            if (desyncDebug_) {
```
I don't see where the logic gets more complicated to put this before the sleep. having it afterwards isn't wrong necessarily but gives more time for the process to get killed by some other means before it actually gets printed.

zdevito commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1255,23 +1257,32 @@ void ProcessGroupNCCL::workCleanupLoop() {
+              collectiveDebugInfoMode_.store(true);
+            }
+
+            c10::optional<std::thread> dumpingDebugInfo;
```
Is this used outside the if below?

wconstab replied:
oh, i did i this way at first so i could preserve having the sleep below the desync clause.

i can put it back that way if it is better. let me see if i can understand your comment above.

fduwjj replied:
Well, I don't know if we sleep before getting the desync report will lead to heartbeat thread kick in early to kill the watchdog so that desync report is not complete.

zdevito reviewed (approved):
Looks good, thanks!

wconstab committed:
Update on "[C10D] Decouple PGNCCL desync from dbg dump"


Add TORCH_NCCL_DUMP_DEBUG_INFO env to control dumping independently
of desync debug feature.

Currently default to disabled (so no behavior change by default),
but plan to default this to true after validation.

Moves 'sleep for 30 sec' that used to be after desync debug to before
it. In my view sleeping before desync is equivalent since we always
sleep the same duration, and keeps the code simpler this way.

Fixes #114433

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 mrshenli zhaojuanmao rohan-varma kiukchung d4l3k lucasllc XilunWu tianyu-l

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Decouple PGNCCL desync from dbg dump"


Add TORCH_NCCL_DUMP_DEBUG_INFO env to control dumping independently
of desync debug feature.

Currently default to disabled (so no behavior change by default),
but plan to default this to true after validation.

Moves 'sleep for 30 sec' that used to be after desync debug to before
it. In my view sleeping before desync is equivalent since we always
sleep the same duration, and keeps the code simpler this way.

Fixes #114433

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 mrshenli zhaojuanmao rohan-varma kiukchung d4l3k lucasllc XilunWu tianyu-l

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Decouple PGNCCL desync from dbg dump"


Add TORCH_NCCL_DUMP_DEBUG_INFO env to control dumping independently
of desync debug feature.

Currently default to disabled (so no behavior change by default),
but plan to default this to true after validation.

Moves 'sleep for 30 sec' that used to be after desync debug to before
it. In my view sleeping before desync is equivalent since we always
sleep the same duration, and keeps the code simpler this way.

Fixes #114433

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 mrshenli zhaojuanmao rohan-varma kiukchung d4l3k lucasllc XilunWu tianyu-l

[ghstack-poisoned]

wconstab commented:
@pytorchbot merge

fduwjj commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -43,6 +43,11 @@ static std::vector<std::string> TORCH_NCCL_ASYNC_ERROR_HANDLING = {
+// Environment Variable to control whether Desync Debug is enabled.
+// This variable must be set together with TORCH_NCCL_ASYNC_ERROR_HANDLING.
+static std::vector<std::string> TORCH_NCCL_DUMP_ON_TIMEOUT = {
+    "TORCH_NCCL_DESYNC_DEBUG"};
```
Wait, are we reusing desync flag for the dump?

wconstab replied:
good catch, thats broken!

wconstab replied:
https://github.com/pytorch/pytorch/pull/114699
